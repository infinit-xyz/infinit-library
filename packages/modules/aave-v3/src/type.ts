import { Address } from 'viem'

export type AaveV3Registry = {
  poolAddressProviderRegistry?: Address
  poolAddressesProvider?: Address
  liquidationLogic?: Address
  supplyLogic?: Address
  eModeLogic?: Address
  flashLoanLogic?: Address
  borrowLogic?: Address
  bridgeLogic?: Address
  poolLogic?: Address
  configuratorLogic?: Address
  reservesSetupHelper?: Address
  poolProxy?: Address
  lendingPools?: Record<string, LendingPool>
  walletBalanceProvider?: Address
  uiIncentiveDataProvider?: Address
  uiPoolDataProviderV3?: Address
  aaveProtocolDataProvider?: Address
  l2PoolImpl?: Address
  poolConfiguratorImpl?: Address
  aclManager?: Address
  aaveOracle?: Address
  aaveEcosystemReserveV2Proxy?: Address
  aaveEcosystemReserveController?: Address
  aaveEcosystemReserveV2Impl?: Address
  reserveInterestRateStrategies?: Record<string, Address>
  wrappedTokenGatewayV3?: Address
  l2Encoder?: Address
  aTokenImpl?: Address
  delegationAwareATokenImpl?: Address
  stableDebtTokenImpl?: Address
  variableDebtTokenImpl?: Address
  poolConfiguratorProxy?: Address
  rewardsControllerImpl?: Address
  rewardsControllerProxy?: Address
  pullRewardsTransferStrategy?: Address
  emissionManager?: Address
  aggregatorApi3Adapters?: Record<string, Address>
  aggregatorPythAdapters?: Record<string, Address>
  aggregatorBandAdapters?: Record<string, Address>
}

export type LendingPool = {
  underlyingToken: Address
  interestRateStrategy: Address
  aToken: Address
  stableDebtToken: Address
  variableDebtToken: Address
}

export type Modify<T, R> = Omit<T, keyof R> & R
