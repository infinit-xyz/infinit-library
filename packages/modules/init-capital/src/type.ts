import { Address } from 'viem'

export type InitCapitalRegistry = {
  proxyAdmin?: Address
  accessControlManager?: Address
  api3ProxyOracleReaderProxy?: Address
  initOracleProxy?: Address
  configProxy?: Address
  liqIncentiveCalculatorProxy?: Address
  posManagerProxy?: Address
  initCoreProxy?: Address
  riskManagerProxy?: Address
  lsdApi3ProxyOracleReaderProxy?: Address
  pythOracleReaderProxy?: Address
  moneyMarketHookProxy?: Address
  initLens?: Address
  api3ProxyOracleReaderImpl?: Address
  initOracleImpl?: Address
  configImpl?: Address
  liqIncentiveCalculatorImpl?: Address
  posManagerImpl?: Address
  initCoreImpl?: Address
  riskManagerImpl?: Address
  lsdApi3ProxyOracleReaderImpl?: Address
  pythOracleReaderImpl?: Address
  moneyMarketHookImpl?: Address
  lendingPoolImpl?: Address
  irms?: Record<string, Address>
  lendingPools?: Record<string, LendingPools>
}

export type LendingPools = {
  underlyingToken: Address
  lendingPool: Address
  irm: Address
}
