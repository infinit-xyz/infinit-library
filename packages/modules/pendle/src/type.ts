import { Address } from 'viem'

export type PendleV3Registry = {
  pendleRouterV4?: Address
  routerStorageV4?: Address
  actionAddRemoveLiqV3?: Address
  actionCallbackV3?: Address
  actionMiscV3?: Address
  actionSimple?: Address
  actionSwapPTV3?: Address
  actionSwapYTV3?: Address
  pendleRouterStatic?: Address
  actionStorageStatic?: Address
  actionInfoStatic?: Address
  actionMarketAuxStatic?: Address
  actionMarketCoreStatic?: Address
  actionMintRedeemStatic?: Address
  actionVePendleStatic?: Address
  pendleLimitRouter?: Address
  pendleLimitRouterProxy?: Address
  proxyAdmin?: Address
  baseSplitCodeFactoryContract?: Address
  oracleLib?: Address
  pendleGaugeControllerMainchainUpg?: Address
  pendleMarketFactoryV3?: Address
  pendlePYLpOracle?: Address
  pendleSwap?: Address
  pendleYieldContractFactory?: Address
  votingEscrowPendleMainchain?: Address
  pendleMsgSendEndpointUpgImpl?: Address
  pendleMsgSendEndpointUpgProxy?: Address
  pendleVotingContollerUpgImpl?: Address
  pendleVotingControllerUpgProxy?: Address
}
