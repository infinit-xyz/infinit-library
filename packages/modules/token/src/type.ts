import { Address, Hex } from 'viem'

export enum TokenType {
  'InfinitERC20' = 'InfinitERC20',
  'InfinitERC20Burnable' = 'InfinitERC20Burnable',
}

export type Token = { type: TokenType }
export type Proxy = { proxyAddress: Address; implementation: Address }

export type TokenRegistry = {
  tokens?: Record<Address, Token>
  tokenDistributors?: Address[]
  accumulativeMerkleDistributors?: Record<Address, Proxy>
  merkleTree?: {
    root: string
    merkle: Record<Address, { amount: string; proof: Hex[] }>
  }
}
