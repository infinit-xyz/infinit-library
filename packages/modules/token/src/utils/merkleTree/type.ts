import { Hex } from 'viem'

export type ProofDetail = {
  proof: Hex[]
  amount: string
}
