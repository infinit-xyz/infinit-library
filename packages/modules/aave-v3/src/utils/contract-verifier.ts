import { Address, PublicClient } from 'viem'

import { BaseContractVerifier, BlockExplorerParams, ContractInfo } from '@infinit-xyz/core/internal'

import { Artifacts } from 'hardhat/types'

import { AaveV3Registry, LendingPool } from '@/src/type'
import { getArtifacts } from '@/src/utils/artifact'

export class AaveV3ContractVerifier extends BaseContractVerifier<AaveV3Registry> {
  constructor(client: PublicClient, params: BlockExplorerParams) {
    super(client, params)
  }

  protected override async getContracts(registry: AaveV3Registry): Promise<ContractInfo[]> {
    const {
      lendingPools,
      reserveInterestRateStrategies,
      poolProxy,
      poolConfiguratorProxy,
      rewardsControllerProxy,
      api3Adapters,
      pythAdapters,
      bandAdapters,
      ...contractToAddressRegistry
    } = registry

    let contracts: ContractInfo[] = []

    for (const address of Object.values(contractToAddressRegistry)) {
      contracts.push({ address })
    }

    if (registry.poolAddressesProvider) {
      if (poolProxy) {
        contracts.push({
          address: poolProxy,
          constructorArgs: [registry.poolAddressesProvider],
        })
      }
      if (poolConfiguratorProxy) {
        contracts.push({
          address: poolConfiguratorProxy,
          constructorArgs: [registry.poolAddressesProvider],
        })
      }
    }

    if (lendingPools && registry.poolConfiguratorProxy) {
      const lendingPoolContractInfos = await this.getLendingPoolContractInfos(registry.poolConfiguratorProxy, lendingPools)
      contracts = contracts.concat(lendingPoolContractInfos)
    }

    if (reserveInterestRateStrategies) {
      const reserveInterestRateStrategyContractInfos = await this.getReserveInterestRateStrategyContractInfos(reserveInterestRateStrategies)
      contracts = contracts.concat(reserveInterestRateStrategyContractInfos)
    }

    if (api3Adapters) {
      const api3AdaptersContractInfos = await this.getAdapterContractInfos(api3Adapters)
      contracts = contracts.concat(api3AdaptersContractInfos)
    }

    if (pythAdapters) {
      const pythAdaptersContractInfos = await this.getAdapterContractInfos(pythAdapters)
      contracts = contracts.concat(pythAdaptersContractInfos)
    }

    if (bandAdapters) {
      const bandAdaptersContractInfos = await this.getAdapterContractInfos(bandAdapters)
      contracts = contracts.concat(bandAdaptersContractInfos)
    }

    return contracts
  }

  protected override async getArtifacts(): Promise<Artifacts> {
    const artifacts = await getArtifacts()

    return artifacts
  }

  private async getLendingPoolContractInfos(poolConfiguratorProxy: Address, lendingPools: Record<string, LendingPool>) {
    const contracts: ContractInfo[] = []
    for (const info of Object.values(lendingPools)) {
      for (const [contractName, address] of Object.entries(info)) {
        if (contractName === 'underlyingToken') continue
        contracts.push({
          address,
          constructorArgs: contractName !== 'interestRateStrategy' ? [poolConfiguratorProxy] : undefined,
        })
      }
    }
    return contracts
  }

  private async getReserveInterestRateStrategyContractInfos(reserveInterestRateStrategies: Record<string, Address>) {
    const contracts: ContractInfo[] = []
    for (const address of Object.values(reserveInterestRateStrategies)) {
      contracts.push({ address })
    }
    return contracts
  }

  private async getAdapterContractInfos(adapters: Record<string, Address>) {
    const contracts: ContractInfo[] = []
    for (const address of Object.values(adapters)) {
      contracts.push({ address })
    }
    return contracts
  }
}
