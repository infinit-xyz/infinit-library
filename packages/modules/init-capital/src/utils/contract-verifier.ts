import { Address, PublicClient } from 'viem'

import { BaseContractVerifier, BlockExplorerParams, ContractInfo } from '@infinit-xyz/core/internal'

import { Artifacts } from 'hardhat/types'

import { InitCapitalRegistry } from '@/src/type'
import { getArtifacts } from '@/src/utils/artifact'

export class TokenContractVerifier extends BaseContractVerifier<InitCapitalRegistry> {
  constructor(client: PublicClient, params: BlockExplorerParams) {
    super(client, params)
  }

  protected override async getContracts(registry: InitCapitalRegistry): Promise<ContractInfo[]> {
    const { lendingPools, irms, ...contractToAddressRegistry } = registry
    const contracts: ContractInfo[] = []

    for (const address of Object.values(contractToAddressRegistry)) {
      contracts.push({ address })
    }

    if (lendingPools) {
      for (const { lendingPool, irm } of Object.values(lendingPools)) {
        contracts.push({ address: lendingPool })
        contracts.push({ address: irm })
      }
    }

    if (irms) {
      for (const address of Object.values(irms)) {
        contracts.push({ address })
      }
    }

    return contracts
  }

  protected override async getArtifacts(): Promise<Artifacts> {
    const artifacts = await getArtifacts()

    return artifacts
  }
}
