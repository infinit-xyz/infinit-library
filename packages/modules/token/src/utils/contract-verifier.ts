import { Address, PublicClient } from 'viem'

import { BaseContractVerifier, BlockExplorerParams, ContractInfo } from '@infinit-xyz/core/internal'

import { Artifacts } from 'hardhat/types'

import { TokenRegistry } from '@/src/type'
import { getArtifacts } from '@/src/utils/artifact'

export class TokenContractVerifier extends BaseContractVerifier<TokenRegistry> {
  constructor(client: PublicClient, params: BlockExplorerParams) {
    super(client, params)
  }

  protected override async getContracts(registry: TokenRegistry): Promise<ContractInfo[]> {
    const { accumulativeMerkleDistributors, tokens } = registry
    const contracts: ContractInfo[] = []

    if (tokens) {
      for (const address of Object.keys(tokens)) {
        contracts.push({ address: address as Address })
      }
    }
    if (accumulativeMerkleDistributors) {
      for (const [proxyAddress, { implementation }] of Object.entries(accumulativeMerkleDistributors)) {
        contracts.push({ address: proxyAddress as Address })
        contracts.push({ address: implementation })
      }
    }

    return contracts
  }

  protected override async getArtifacts(): Promise<Artifacts> {
    const artifacts = await getArtifacts()

    return artifacts
  }
}
