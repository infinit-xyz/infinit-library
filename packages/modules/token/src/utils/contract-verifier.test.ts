import { describe, expect, test } from 'vitest'

import { createPublicClient, http } from 'viem'
import { holesky } from 'viem/chains'

import { TokenRegistry, TokenType } from '../type'
import { TokenContractVerifier } from './contract-verifier'

// Create a subclass to expose the protected method for testing
class TestTokenContractVerifier extends TokenContractVerifier {
  public async testGetContracts(registry: TokenRegistry) {
    return this.getContracts(registry)
  }
}

describe('TokenContractVerifier.ts', () => {
  const registry: TokenRegistry = {
    tokens: {
      '0x86ce7353a19a6dbdb0a0ac01536ffc60f704980e': {
        type: TokenType.InfinitERC20,
      },
      '0x445aa85126dd02145e194e31b19ebac0d051e4e2': {
        type: TokenType.InfinitERC20Burnable,
      },
    },
    accumulativeMerkleDistributors: {
      '0xab7819826b1acd87e30fa45e359e4038683365e6': {
        implementation: '0x78bb7fd5fc49a644f7b07e43812ef9535fbcfd29',
      },
      '0x63743aaf4522ad8c47ba84b2fd2f645710b4454a': {
        implementation: '0x56313c36ba155bab1a5d569f48eaa965e270399d',
      },
    },
  }

  test('getContracts: should match snapshot', async () => {
    const verifier = new TestTokenContractVerifier(
      await createPublicClient({
        chain: holesky,
        transport: http('https://1rpc.io/holesky'),
      }),
      {
        apiKey: process.env.HOLESKY_ETHERSCAN_API_KEY ?? '',
        apiUrl: 'https://api-holesky.etherscan.io/api',
        url: 'https://holesky.etherscan.io',
      },
    )
    expect(verifier.testGetContracts(registry)).resolves.toMatchSnapshot()
  })
})
