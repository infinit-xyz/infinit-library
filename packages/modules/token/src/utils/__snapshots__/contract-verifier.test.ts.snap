// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TokenContractVerifier.ts > export: should match snapshot 1`] = `
[
  {
    "address": "0x86ce7353a19a6dbdb0a0ac01536ffc60f704980e",
  },
  {
    "address": "0x445aa85126dd02145e194e31b19ebac0d051e4e2",
  },
  {
    "address": "0xab7819826b1acd87e30fa45e359e4038683365e6",
  },
  {
    "address": "0x78bb7fd5fc49a644f7b07e43812ef9535fbcfd29",
  },
  {
    "address": "0x63743aaf4522ad8c47ba84b2fd2f645710b4454a",
  },
  {
    "address": "0x56313c36ba155bab1a5d569f48eaa965e270399d",
  },
]
`;

exports[`TokenContractVerifier.ts > getContracts: should match snapshot 1`] = `
[
  {
    "address": "0x86ce7353a19a6dbdb0a0ac01536ffc60f704980e",
  },
  {
    "address": "0x445aa85126dd02145e194e31b19ebac0d051e4e2",
  },
  {
    "address": "0xab7819826b1acd87e30fa45e359e4038683365e6",
  },
  {
    "address": "0x78bb7fd5fc49a644f7b07e43812ef9535fbcfd29",
  },
  {
    "address": "0x63743aaf4522ad8c47ba84b2fd2f645710b4454a",
  },
  {
    "address": "0x56313c36ba155bab1a5d569f48eaa965e270399d",
  },
]
`;
